File: julia/ch1/ex108.jl
#=
ex108:
- Julia version: 0.6.4
- Author: Yohan
- Date: 2018-08-01
=#

function square(x)
  x * x
end

function cube(x)
  x * x * x
end

function improve(guess, x)
  (x / square(guess) + 2 * guess) / 3.0
end

function good_enough(guess, x)
  abs(cube(guess) - x) < (guess / 100000.0)
end

function cbrt_iter(guess, x)
  good_enough(guess, x) ? guess : cbrt_iter(improve(guess, x), x)
end

function my_cbrt(x)
  cbrt_iter(1.0, x)
end

File: erlang/ch1/ex108.erl
%%%-------------------------------------------------------------------
%%% @author Yohan
%%% @copyright (C) 2018, <COMPANY>
%%% @doc
%%% compile with > c("ex108.erl").
%%% load with > l(ex108).
%%% run with > ex108:cbrt(x).
%%% @end
%%% Created : 27. Jul 2018 2:36 PM
%%%-------------------------------------------------------------------
-module(ex108).
-author("Yohan").

%% API
-export([cbrt/1]).


square(X) -> X * X.

cube(X) -> X * X * X.

improve(Guess, X) -> (X / square(Guess) + 2 * Guess) / 3.0.

good_enough(Guess, X) -> abs(cube(Guess) - X) < (Guess / 100000.0).

cbrt_iter(Guess, X) ->
  case good_enough(Guess, X) of
    true -> Guess;
    false -> cbrt_iter(improve(Guess, X), X)
  end.

cbrt(X) -> cbrt_iter(1.0, X).
